src/thunderbots/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,200 @@
+    // Take ownership of the in_destructor flag so we can use it for the conditional
+    // check
+    in_destructor_mutex.lock();
+    while(!in_destructor){
+        // Give up ownership of the in_destructor flag now that we're done the
+        // conditional check
+        in_destructor_mutex.unlock();
+
+        optimizePasses();
+        pruneAndReplacePasses();
+        saveBestPath();
+
+        // Yield to allow other threads to run. This is particularly important if we
+        // have this thread and another running on one core
+        std::this_thread::yield();
+
+        // Take ownership of the `in_destructor` flag so we can use it for the conditional
+        // check
+        in_destructor_mutex.lock();
+    }
+}
+
+void PassGenerator::optimizePasses() {
+    // The objective function we minimize in gradient descent to improve each pass
+    // that we're optimizing
+    static const auto objective_function =
+            [this](std::array<double, NUM_PARAMS_TO_OPTIMIZE> pass_array){
+                Pass pass = convertArrayToPass(pass_array);
+                return ratePass(pass);
+            };
+
+    // Run gradient descent to optimize the passes to for the requested number
+    // of iterations
+    // TODO: Can we parallize this? Would it make it faster?
+    for (Pass& pass : passes_to_optimize){
+        auto pass_array = optimizer.maximize(
+                objective_function,
+                convertPassToArray(pass),
+                number_of_gradient_descent_steps_per_iter
+        );
+        pass = convertArrayToPass(pass_array);
+    }
+}
+
+void PassGenerator::pruneAndReplacePasses() {
+    // Sort the passes by quality
+    std::sort(passes_to_optimize.begin(), passes_to_optimize.end(),
+              [this](Pass p1, Pass p2) { return comparePassQuality(p1, p2); }
+    );
+
+    // Replace the least promising passes with newly generated passes
+    if (num_passes_to_keep_after_pruning < num_passes_to_optimize){
+        std::vector<Pass> new_passes = generatePasses(num_passes_to_optimize -
+                                                      num_passes_to_keep_after_pruning);
+
+        // Remove the worst paths
+        if (num_passes_to_keep_after_pruning < passes_to_optimize.size()) {
+            passes_to_optimize.erase(passes_to_optimize.begin() + num_passes_to_keep_after_pruning,
+                    passes_to_optimize.end());
+        }
+
+        // Append our newly generated passes to replace the passes we just removed
+        passes_to_optimize.insert(
+                passes_to_optimize.begin() + num_passes_to_keep_after_pruning,
+                new_passes.begin(), new_passes.end());
+    }
+
+}
+
+void PassGenerator::saveBestPath() {
+    // Take ownership of the best_known_pass for the duration of this function
+    std::lock_guard<std::mutex> best_known_pass_lock(best_known_pass_mutex);
+
+    std::sort(passes_to_optimize.begin(), passes_to_optimize.end(),
```

Comment:
Again, we should specify if we are sorting in increasing or decreasing order.

Reply:
added